internal/idle/idle.go

```diff
@@ -203,22 +176,46 @@ func (m *manager) tryEnterIdleMode() bool {
 	// No new RPCs have come in since we last set the active calls count value
 	// -math.MaxInt32 in the timer callback. And since we have the lock, it is
 	// safe to enter idle mode now.
-	if err := m.enforcer.EnterIdleMode(); err != nil {
-		m.logger.Errorf("Failed to enter idle mode: %v", err)
-		return false
-	}
+	m.enforcer.EnterIdleMode()
 
 	// Successfully entered idle mode.
 	m.actuallyIdle = true
 	return true
 }
 
+func (m *Manager) EnterIdleModeForTesting() {
+	if m.timeout == 0 {
+		m.idleMu.Lock()
+		defer m.idleMu.Unlock()
+		if !m.actuallyIdle {
+			m.enforcer.EnterIdleMode()
+		}
+		return
+	}
+	if !atomic.CompareAndSwapInt32(&m.activeCallsCount, 0, -math.MaxInt32) {
+		// We have an active RPC and cannot enter idle mode.
+		return
+	}
+	if m.tryEnterIdleMode() {
+		// Successfully entered idle mode. No further action necessary.
+		return
+	}
+	// Failed to enter idle mode due to a concurrent RPC that kept the channel
+	// active, or because of an error from the channel. Undo the attempt to
+	// enter idle, and reset the timer to try again later.
+	atomic.AddInt32(&m.activeCallsCount, math.MaxInt32)
+}
+
 // OnCallBegin is invoked at the start of every RPC.
-func (m *manager) OnCallBegin() error {
+func (m *Manager) OnCallBegin() error {
 	if m.isClosed() {
 		return nil
 	}
-
+	if m.timeout == 0 {
+		// When the manager is disabled (timeout==0), we just exit idle mode if
+		// needed and return.
+		return m.ExitIdleMode()
+	}
```

Comment:
If idleness is disabled, this means that for every RPC, we would be taking the lock (in `m.ExitIdleMode()`), which seems like a lot of wasted effort and a potential for lock contention issues in environments with a high QPS.

Reply:
Hmm, good point.  So should we just add back in all the accounting for the disabled idleness manager?

And this means we take this lock on every RPC except when there is a concurrent call already running.  I wonder if we can improve this in the base case, too?

Reply:
I just removed all short circuiting of the accounting for m.timeout==0.  We can save any other improvements for later.